packages/assets-controllers/src/DeFiPositionsController/DeFiPositionsController.ts

```diff
@@ -0,0 +1,245 @@
+  AccountsControllerListAccountsAction;
+
+/**
+ * The external events available to the {@link DeFiPositionsController}.
+ */
+export type DeFiPositionsControllerAllowedEvents =
+  | KeyringControllerUnlockEvent
+  | KeyringControllerLockEvent
+  | TransactionControllerTransactionConfirmedEvent
+  | AccountsControllerAccountAddedEvent;
+
+/**
+ * The messenger of the {@link DeFiPositionsController}.
+ */
+export type DeFiPositionsControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  DeFiPositionsControllerActions | DeFiPositionsControllerAllowedActions,
+  DeFiPositionsControllerEvents | DeFiPositionsControllerAllowedEvents,
+  DeFiPositionsControllerAllowedActions['type'],
+  DeFiPositionsControllerAllowedEvents['type']
+>;
+
+/**
+ * Controller that stores assets and exposes convenience methods
+ */
+export class DeFiPositionsController extends StaticIntervalPollingController()<
+  typeof controllerName,
+  DeFiPositionsControllerState,
+  DeFiPositionsControllerMessenger
+> {
+  readonly #fetchPositions: (
+    accountAddress: string,
+  ) => Promise<DefiPositionResponse[]>;
+
+  readonly #isEnabled: () => boolean;
+
+  /**
+   * Tokens controller options
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - The controller messenger.
+   * @param options.isEnabled - Function that returns whether the controller is enabled. (default: () => true)
+   */
+  constructor({
+    messenger,
+    isEnabled = () => true,
+  }: {
+    messenger: DeFiPositionsControllerMessenger;
+    isEnabled?: () => boolean;
+  }) {
+    super({
+      name: controllerName,
+      metadata: controllerMetadata,
+      messenger,
+      state: getDefaultDefiPositionsControllerState(),
+    });
+
+    this.setIntervalLength(TEN_MINUTES_IN_MS);
+
+    this.#fetchPositions = buildPositionFetcher();
+    this.#isEnabled = isEnabled;
+
+    this.messagingSystem.subscribe('KeyringController:unlock', async () => {
+      this.startPolling(null);
+    });
+
+    this.messagingSystem.subscribe('KeyringController:lock', () => {
+      this.stopAllPolling();
+    });
+
+    this.messagingSystem.subscribe(
+      'TransactionController:transactionConfirmed',
+      async (transactionMeta) => {
+        if (!this.#isEnabled()) {
+          return;
+        }
+
+        await this.#updateAccountPositions(transactionMeta.txParams.from);
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAdded',
+      async (account) => {
+        if (!this.#isEnabled() || !account.type.startsWith('eip155:')) {
+          return;
+        }
+
+        await this.#updateAccountPositions(account.address);
+      },
+    );
+  }
+
+  async _executePoll(): Promise<void> {
+    if (!this.#isEnabled()) {
+      return;
+    }
+
+    const accounts = this.messagingSystem.call(
+      'AccountsController:listAccounts',
+    );
+
+    const results = await reduceInBatchesSerially({
```

Comment:
I found this handy function lying around. I would have preferred something like p-queue, which doesn't do batches but limits how many promises execute at once, but really not worth adding a new dependency for that.

We set a healthy limit of 10 addresses at a time, we might have to increase or reduce it.